modules/mod-deep-learning/DeepLearningEffectBase.cpp

```diff
@@ -0,0 +1,363 @@
+   {
+      sampleCount clipStart = clip->GetPlayStartSample();
+      sampleCount clipEnd = clip->GetPlayEndSample();
+
+      // skip if clip is out of bounds
+      if (start > clipEnd || end < clipStart)
+         continue;
+
+      // trim around the edges
+      start > clipStart ? 
+         clipStart = start : clipStart;
+
+      end < clipEnd ?
+         clipEnd = end : clipEnd;
+
+      sampleCount samplePos = clipStart;
+
+      while (samplePos < clipEnd)
+      {
+         //Get a blockSize of samples (smaller than the size of the buffer)
+         size_t blockSize = limitSampleBufferSize(
+            track->GetBestBlockSize(samplePos),
+            clipEnd - samplePos);
+         
+         blockIndices.emplace_back(BlockIndex(samplePos, blockSize));
+
+         samplePos += blockSize;
+      }
+   }
+
+   return blockIndices;
+}
+
+std::vector<ClipTimestamps> DeepLearningEffectBase::GetClipTimestamps(WaveTrack *track, double tStart, double tEnd) const
+{
+   std::vector<ClipTimestamps> timestamps;
+
+   for (const auto &clip : track->GetClips())
+      timestamps.emplace_back(
+         std::pair<double, double>(clip->GetPlayStartTime(), 
+                                   clip->GetPlayEndTime())
+      );
+
+   return timestamps;
+}
+
+
+torch::Tensor DeepLearningEffectBase::BuildMonoTensor(WaveTrack *track, float *buffer,
+                                                  sampleCount start, size_t len)
+{
+   //Get the samples from the track and put them in the buffer
+   if (!track->GetFloats(buffer, start, len))
+      throw std::runtime_error("An error occurred while copying samples to tensor buffer.");
+
+   // get tensor input from buffer
+   torch::Tensor audio = torch::from_blob(buffer, len,
+                                          torch::TensorOptions().dtype(torch::kFloat));
+   audio = audio.unsqueeze(0); // add channel dimension
+
+   return audio;
+}
+
+torch::Tensor DeepLearningEffectBase::BuildMultichannelTensor(WaveTrack *leader, float *buffer,
+                                                          sampleCount start, size_t len)
+{
+   auto channels = TrackList::Channels(leader);
+   std::vector<torch::Tensor> channelStack;
+
+   // because we're reusing the same buffer, it's important that
+   // we clone the tensor.
+   for (WaveTrack *channel : channels)
+      channelStack.emplace_back(
+          BuildMonoTensor(channel, buffer, start, len).clone());
+
+   return torch::cat(channelStack);
+}
+
+torch::jit::IValue DeepLearningEffectBase::ForwardPassInThread(torch::Tensor input)
+{
+   torch::jit::IValue output;
+
+   std::atomic<bool> done = {false};
+   std::atomic<bool> success = {true};
+
+   // make a copy of the model (in case we need to abort)
+   DeepModelHolder model = this->mModel;
+
+   auto thread = std::thread(
+      [model, &input, &output, &done, &success]()
+      {
+         try
+         {
+            torch::jit::IValue tempOut = model->Forward(input);
```

Comment:
Does the library provide no means to execute `forward` incrementally and interruptible?  That is unfortunate if so, and would explain some of what I am reading.

Or do I understand that there is such a means, and you did use it at the higher level, where the effect and the analyzer iterate over blocks of the track?

So you are only adding precautions here against a model that still gets hung doing a small piece?

Should you explain that in comments?

Reply:
libtorch does not provide a way to execute `forward` incrementally and interruptibly. I'm processing the audio clip-by-clip. Each clip is processed in blocks, where I use `track->GetBestBlockSize` to get the size of each block. I find that the blocks are in the order of seconds, but deep learning models can still take a while when processing a few seconds of audio. @crsib suggested to not do `forward` in the main thread, because of how slow some deep models are to process audio.